libraries/account-resolution/README.md

```diff
@@ -0,0 +1,130 @@
+Since it's impossible to load additional accounts on-chain, if a program requires
+additional accounts to properly implement an instruction, there's no clear way
+for clients to fetch these accounts.
+
+There are two main ways to fetch additional accounts, on-chain and off-chain.
+
+### Inside Program Logic
+
+To expose the additional accounts required, instruction interfaces can include
+supplemental instructions that just give the accounts required.
+
+For example, a `Transferable` interface, along with a `transfer` instruction,
+also requires implementations to expose a `get_additional_accounts_for_transfer`
+instruction.
+
+In the program implementation, this instruction writes the additional accounts
+into return data, making it easy for on-chain and off-chain clients to consume.
+
+See this
+[relevant sRFC](https://forum.solana.com/t/srfc-00010-additional-accounts-request-transfer-spec/122)
+for more information.
+
+### Outside Program Logic
+
+It's also possible for programs to write the additional required account infos
+into account data, so that on-chain and off-chain clients simply need to read
+the data to figure out the additional required accounts.
+
+Rather than exposing this data dynamically through program execution, this method
+uses static account data.
+
+For the `Transferable` interface example, rather than requiring programs to
+implement `get_additional_accounts_for_transfer`, this approach requires programs
+to write to an account defined at a given address. This could be directly in the
+`mint`, or some address derivable from the mint address.
+
+Off-chain, a client must fetch this additional account and read its data to find
+out the additional required accounts, and then include them in the instruction.
+
+On-chain, a program must have access to "remaining account infos" containing the
+special account and all other required accounts to properly create the CPI
+instruction and give the correct account infos.
+
+This could be called a "state interface".
+
+## How it works
+
+This library uses `spl-type-length-value` to read and write required instruction
+accounts from account data.
+
+Interface instructions must have an 8-byte discriminator, so that the exposed
+`ExtraAccountMetas` type can use the instruction discriminator as a `TlvDiscriminator`.
+
+This can be confusing. Typically, a type implements `TlvDiscriminator`, so that
+the type can be written into TLV data. In this case, `ExtraAccountMetas` is
+generic over `TlvDiscriminator`, meaning that a program can write many different instances of
+`ExtraAccountMetas` into one account, using different `TlvDiscriminator`s.
+
+Also, it's reusing an instruction discriminator as a TLV discriminator. For example,
+if the `transfer` instruction has a discriminator of `[1, 2, 3, 4, 5, 6, 7, 8]`,
+then the account uses a TLV discriminator of `[1, 2, 3, 4, 5, 6, 7, 8]` to denote
+where the additional account metas are stored.
```

Comment:
This is really cool. Looking forward to seeing examples of managing this state on-chain.